### PR TITLE
[kernel][moe] Optimize moeTopK with top-2-per-iteration fast path

### PR DESCRIPTION
**Pull request by fanlei23**
## Purpose

The default `moeTopK` path (non-power-of-2 or >512 experts) picks one expert per iteration with an O(k) inner loop to skip prior winners. This is fine for small k but gets slow for models like DeepSeek-V3 (256 experts, topk=8).

Two changes:
- New `moeTopKFast` kernel: picks top-2 per iteration via a `TopKPair` CUB reducer, cutting iterations from k to ceil(k/2). Dispatched when topk > 1.
- Both kernels now mark selected scores in-place (`= -1.f`) instead of the inner loop re-scan. Eliminates the O(k^2) comparison.

The fused `topkGating` path (power-of-2 experts <= 512) is unchanged.

## Test Plan

No model weights needed. All tests use random gating outputs on GPU.

738 configs covering num_tokens {1,16,128,512,1024,2048}, num_experts {4..512}, topk {1..8}, dtype {fp16,bf16,fp32}, renormalize {true,false}. Each config compares kernel output against `torch.softmax + torch.topk` reference.

```bash
pytest tests/kernels/moe/test_topk_fast.py -v
```

Tested on H200 (sm_90), CUDA 12.x, PyTorch 2.6.

## Test Result

Correctness -- all 738 cases pass (before and after):

```
test_topkfast_softmax (512 experts, topk 1-8):  36/36
test_topk_softmax (4-256 experts, topk 1-4):   126/126
test_topk_softmax_dtype (fp16/bf16/fp32):       432/432
test_topk_softmax_renormalize:                  144/144
```

Performance on the default path (H200, non-power-of-2 experts):

```
  tokens  experts  topk   before(ms)    after(ms)   speedup
    1024      257     8       0.0291       0.0241     1.20x
    2048      257     4       0.0286       0.0253     1.13x
    2048      257     8       0.0472       0.0372     1.27x
    2048      500     4       0.0331       0.0278     1.19x
    2048      500     8       0.0567       0.0404     1.40x
    4096      257     4       0.0474       0.0412     1.15x
    4096      257     8       0.0825       0.0636     1.30x
```

Speedup scales with k: ~1.1x at topk=2/4, up to 1.4x at topk=8. Fused path unaffected.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: csrc/moe/topk_softmax_kernels.cu [hunk 1]
```diff
@@ -156,9 +156,131 @@ __launch_bounds__(TPB) __global__
...
+                const int idx = k * block_row + k_idx * 2 + i;
+                output[idx] = inputs_after_softmax[thread_read_offset + expert];
+                indices[idx] = should_process_row ? (expert - start_expert) : num_experts;
+                assert(indices[idx] >= 0);
+                source_rows[idx] = (k_idx * 2 + i) * num_rows + block_row;
+                if (renormalize) {
+                    selected_sum += output[idx];
+                }
+                inputs_after_softmax[thread_read_offset + expert] = -1.f;
+            }
```
[the remark this answers is not in the document]
Comment (fanlei23) [reply csrc/moe/topk_softmax_kernels.cu:261]: Under normal MoE usage (top-2/4/8 selection, far fewer than total experts), valid experts are always available, so no invalid or duplicate expert 0 will be selected, and -1.f will never be accumulated into selected_sum.
The original -1.f sentinel is safe in practice because softmax outputs with bias do not reach this value in real workloads, and only previously selected experts are marked with it.

File: csrc/moe/topk_softmax_kernels.cu [hunk 2]
```diff
@@ -156,9 +156,131 @@ __launch_bounds__(TPB) __global__
...
+            globalMax = b.max;
+        }
+
+        if (globalMax.key == a.max.key) {
+            globalSecondMax = (a.secondMax.value > b.max.value) ? a.secondMax : b.max;
+        } else {
+            globalSecondMax = (b.secondMax.value > a.max.value) ? b.secondMax : a.max;
+        }
+        return TopKPair(globalMax, globalSecondMax);
+    }
```
[the remark this answers is not in the document]
Comment (fanlei23) [reply csrc/moe/topk_softmax_kernels.cu:183]: good job. Fix it

**Comment by fanlei23** [comment]
cc@yewentao256 @WoosukKwon

**Comment by fanlei23** [comment]
cc @zyongye @mgoin @tlrmchlsmth
